src/Task/ProtectedTaskManager.cpp

```diff
@@ -84,29 +86,50 @@ ProtectedTaskManager::IncrementActiveTaskPointArm(int offset)
 {
   ExclusiveLease lease(*this);
   TaskAdvance &advance = lease->SetTaskAdvance();
+  TaskWaypoint *twp = nullptr;
 
   switch (advance.GetState()) {
   case TaskAdvance::MANUAL:
-  case TaskAdvance::AUTO:
-    lease->IncrementActiveTaskPoint(offset);
+  case TaskAdvance::AUTO: {
+      lease->IncrementActiveTaskPoint(offset);
+      twp = lease->GetActiveTaskPoint();
+    }
     break;
   case TaskAdvance::START_DISARMED:
   case TaskAdvance::TURN_DISARMED:
     if (offset>0) {
       advance.SetArmed(true);
     } else {
       lease->IncrementActiveTaskPoint(offset);
+      twp = lease->GetActiveTaskPoint();
     }
     break;
   case TaskAdvance::START_ARMED:
   case TaskAdvance::TURN_ARMED:
     if (offset>0) {
       lease->IncrementActiveTaskPoint(offset);
+      twp = lease->GetActiveTaskPoint();
     } else {
       advance.SetArmed(false);
     }
     break;
   }
+  if (twp) {
+    switch (twp->GetType()) {
+    case TaskPointType::AST : {
+      ASTPoint *p = (ASTPoint *)twp;
+      if (p->HasEntered()) p->Reset();
+      }
+      break;
+    case TaskPointType::AAT : {
+      AATPoint *p = (AATPoint *)twp;
+      if (p->HasEntered()) p->Reset();
+      }
```

Comment:
Duplicate code.
Why check `HasEntered()`?

Reply:
Duplicate code: Can I assume classes ASTPoint and AATPoint would do the same???? I didn't want to make that assumption.

Agreed: I don't have to check HasEntered(). But doesn't it improve the readability?

Reply:
> Can I assume classes ASTPoint and AATPoint would do the same?

They don't do the same. But you do the same with both types. And that's what your code is about: what you do *with* those objects, not what *they* do.

> Agreed: I don't have to check HasEntered(). But doesn't it improve the readability?

Does it? Unnecessary checks don't improve readability. They reduce readability by having useless code, and they reduce performance.

Maybe the check is necessary - I did not check - I was just asking what was your intention why you wrote it.